### PR TITLE
Support what RubyGems needs

### PR DESCRIPTION
**Pull request by segiddins**

### CODE REVIEW

File: lib/molinillo/resolution.rb [hunk 1]
```diff
@@ -394,19 +394,25 @@ def require_nested_dependencies_for(activated_spec)
...
+        handle_missing_or_push_dependency_state DependencyState.new(
+          new_name, new_requirements, new_activated,
+          new_requirement, possibilities, depth, conflicts.dup
         )
       end
+
+      def handle_missing_or_push_dependency_state(state)
+        if state.requirement && state.possibilities.empty? && allow_missing?(state.requirement)
+          state.activated.detach_vertex_named(state.name)
+          return push_state_for_requirements(state.requirements, state.activated)
```
Comment (Who828) [inline lib/molinillo/resolution.rb]: no need to return explicitly with an else block.

File: lib/molinillo/modules/specification_provider.rb [hunk 2]
```diff
@@ -26,7 +26,7 @@ def search_for(dependency)
     # @param [Object] specification
     # @return [Array<Object>] the dependencies that are required by the given
     #   `specification`.
-    def dependencies_for(specification)
+    def dependencies_for(specification, activated)
```
Comment (segiddins) [inline lib/molinillo/modules/specification_provider.rb]: this no longer needs to take the DependencyGraph parameter

**Comment by segiddins** [comment]
This needs some documentation and some testing around when `allow_missing?` returns true

File: lib/molinillo/modules/specification_provider.rb (hunk 2 above)
Comment (alloy) [inline lib/molinillo/modules/specification_provider.rb]: So what’s `activated`? (Can you add docs for it please?)

Comment (segiddins) [inline lib/molinillo/modules/specification_provider.rb]: haha, that's the parameter I can remove now :)

**Comment by alloy** [comment]
Off-topic, but probably a good idea to keep things like this in mind, if your work will impact that at all https://github.com/rubygems/rubygems/pull/1188.

File: lib/molinillo/modules/specification_provider.rb (hunk 2 above)
Comment (alloy) [inline lib/molinillo/modules/specification_provider.rb]: Ah, hah, even better :D

**Comment by segiddins** [comment]
OK, I'm going to merge this tomorrow and cut a release unless there are any objections (\c @indirect).

**Comment by indirect** [comment]
:sheep: it

**Comment by segiddins** [comment]
:shipit:

**Pull request merged**